raulin0/backup-juggler
# feat: adds backup feature

## Description
This pull request introduces comprehensive documentation for the Backup Juggler application. It covers installation instructions, a quick start guide, usage tutorials, contribution guidelines, and API documentation. Additionally, it includes updates to the project dependencies, adds a script entry, and introduces documentation configuration. Furthermore, it adds additional tests for the Backup Juggler CLI and unit tests for the backup functionality, along with necessary fixtures. Lastly, it adds Backup Juggler CLI functionality, including the ProgressViewer class, Paths class, and file backup implementation.

## Changes Made
### Documentation
- Added installation documentation.
- Added quick start documentation.
- Added usage tutorial documentation.
- Added contribution documentation.
- Added Backup Juggler API documentation.
- Added CSS configuration for the cards from the quick start documentation.
- Added images for the tutorial usage documentation.
### Build
- Updated project dependencies.
- Added script entry.
- Introduced documentation configuration.
### Test
- Added several new test cases for the Backup Juggler CLI. The tests cover running the 'bj' script successfully, printing version and help messages, performing backups, and calculating the size of sources. The tests utilize fixtures such as 'run_backup_juggler', 'create_files', 'create_directories', and 'create_subdirectories_recursively' for test setup and invoke the CLI commands using 'subprocess.run' and 'CliRunner' from the 'typer.testing' module. These tests enhance the test coverage for the Backup Juggler CLI.
- Added the 'run_backup_juggler' fixture to facilitate testing of the Backup Juggler CLI. It allows specifying options, subcommands, sources, and destinations as parameters and invokes the CLI command using 'CliRunner' from the 'typer.testing' module. This fixture will assist in writing test cases for the Backup Juggler CLI.
- Added unit tests for the backup functionality in the Backup Juggler application. The tests cover the functionality of the 'backup' and 'parallel_backups' methods in the 'files_manager' module, as well as various scenarios for file and directory copying. The tests utilize parameterization to test different scenarios and utilize test fixtures to create the necessary file and directory structures for testing.
- Added fixtures to create temporary paths in the Backup Juggler application. The fixtures include 'create_files', 'create_directories', and 'create_subdirectories_recursively' to facilitate the creation of files, directories, and directory structures for testing purposes. These fixtures provide convenient methods for generating test data during unit tests and enhance the flexibility and effectiveness of testing within the Backup Juggler application.
### Feature
- Added the ProgressViewer class, responsible for displaying the progress of the file copying operation using a progress bar. The class utilizes the tqdm library to create a progress bar that visualizes the progress of the file copying operation. The _update method is implemented to update the progress bar with the given chunk size.
- Added the Paths class, which represents the source and destination paths in the Backup Juggler application. The class provides methods to retrieve the source path, destination directory, and calculate the total size of the files to be copied.
- Implemented the file backup functionality in the Backup Juggler application. The code includes the following components:
- FileCopier class: This class acts as the controller for the file copying process. It provides methods for copying files from a source directory to a destination directory while preserving the directory structure. The class utilizes the Paths and ProgressViewer instances to manage source and destination paths and display progress.
- calculates_size function: This function is responsible for calculating the total size of the provided source directories or files. It iterates through the list of source paths, accumulates the sizes of each source, and displays the result in a human-readable format.
- backup function: This function initiates a single backup operation from a source directory or file to a destination directory. It creates instances of the Paths, ProgressViewer, and FileCopier classes and invokes the _copy_to method of the FileCopier instance to perform the actual file copying.
- parallel_backups function: This function enables the execution of multiple file backup operations in parallel. It utilizes the ThreadPoolExecutor from the concurrent.futures module to submit backup tasks for each combination of source and destination paths. The function waits for all tasks to complete before returning.

## Motivation and Context
The documentation aims to provide a comprehensive guide for users to understand and utilize the Backup Juggler application effectively. The tests and fixtures enhance the test coverage and reliability of the Backup Juggler CLI and backup functionality. The new features, including the ProgressViewer class, Paths class, and file backup implementation, improve the functionality and usability of the Backup Juggler application.

## How to Test
1. Follow the installation documentation to set up the Backup Juggler application.
2. Refer to the quick start documentation for a basic overview of the application's usage.
3. Explore the usage tutorial documentation to understand the advanced features and capabilities of the application.
4. Read the contribution documentation to learn how to contribute to the project.
5. Check the Backup Juggler API documentation for detailed information on the available classes and methods.
6. Verify that the project dependencies are updated and the script entry is added.
7. Execute the additional tests for the Backup Juggler CLI and unit tests for the backup functionality. Ensure that the fixtures are used appropriately for test setup and invocation of CLI commands.
8. Utilize the created fixtures, such as 'create_files', 'create_directories', and 'create_subdirectories_recursively', to generate temporary paths for unit tests.
9. Validate the functionality of the Backup Juggler CLI, including successful script execution, correct output of version and help messages, backup operations, and accurate calculation of source sizes.
10. Test the file backup functionality by performing backup operations from various source directories or files to destination directories. Ensure that the progress bar is displayed correctly, and files are copied with the appropriate directory structure.
11. Execute parallel backup operations and verify that they complete successfully, using multiple combinations of source and destination paths.

## Related Issue(s)
N/A

## Checklist
- [x] Documentation has been added or updated to reflect the new functionality.

Please review and merge this pull request at your earliest convenience. Feel free to provide any feedback or suggestions for improvement. Thank you!
